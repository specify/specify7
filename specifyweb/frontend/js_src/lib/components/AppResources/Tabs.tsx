import { Tab } from '@headlessui/react';
import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { resourcesText } from '../../localization/resources';
import { f } from '../../utils/functools';
import type { GetSet, IR, RA, RR } from '../../utils/types';
import { filterArray } from '../../utils/types';
import { WarningMessage } from '../Atoms';
import { toResource } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { getResourceApiUrl } from '../DataModel/resource';
import { schema } from '../DataModel/schema';
import type {
  SpAppResource,
  SpAppResourceDir,
  SpViewSetObj as SpViewSetObject,
} from '../DataModel/types';
import { ErrorBoundary } from '../Errors/ErrorBoundary';
import { Dialog, dialogClassNames } from '../Molecules/Dialog';
import { appResourceIcon } from './EditorComponents';
import { radioButtonClassName } from './Filters';
import { getAppResourceType, getResourceType } from './filtersHelpers';
import type {
  AppResourceEditorType,
  AppResourceTabProps,
} from './TabDefinitions';
import {
  AppResourceTextEditor,
  visualAppResourceEditors,
} from './TabDefinitions';

export function AppResourcesTab({
  tab: Component,
  label,
  headerButtons,
  appResource,
  resource,
  directory,
  data,
  isFullScreen: [isFullScreen, _],
  onChange: handleChange,
  onSetCleanup: setCleanup,
  footer,
}: {
  readonly tab: Component;
  readonly label: LocalizedString;
  readonly appResource: SpecifyResource<SpAppResource | SpViewSetObject>;
  readonly resource: SerializedResource<SpAppResource | SpViewSetObject>;
  readonly directory: SerializedResource<SpAppResourceDir>;
  readonly headerButtons: JSX.Element;
  readonly data: string | null;
  readonly isFullScreen: GetSet<boolean>;
  readonly onChange: (
    data: string | (() => string | null | undefined) | null
  ) => void;
  readonly onSetCleanup: (callback: () => Promise<void>) => void;
  readonly footer: JSX.Element;
}): JSX.Element {
  const children = (
    <ErrorBoundary dismissible>
      <Component
        appResource={appResource}
        data={data}
        directory={directory}
        resource={resource}
        onChange={handleChange}
        onSetCleanup={setCleanup}
      />
    </ErrorBoundary>
  );
  return isFullScreen ? (
    <Dialog
      buttons={<div className="flex gap-2">{footer}</div>}
      className={{
        container: dialogClassNames.fullScreen,
        buttonContainer: '!block',
      }}
      dimensionsKey={false}
      header={label}
      headerButtons={headerButtons}
      icon={appResourceIcon(getResourceType(resource))}
      onClose={undefined}
    >
      {children}
    </Dialog>
  ) : (
    children
  );
}

type Component = (props: AppResourceTabProps) => JSX.Element;

export function useEditorTabs(
  resource: SerializedResource<SpAppResource | SpViewSetObject>
): RA<{
  readonly label: LocalizedString;
  readonly component: (props: AppResourceTabProps) => JSX.Element;
}> {
  const subType =
    f.maybe(toResource(resource, 'SpAppResource'), getAppResourceType) ??
    'viewSet';
  return React.useMemo(() => {
    const editors =
      typeof subType === 'string'
        ? visualAppResourceEditors()[subType]
        : undefined;
    return editors === undefined
      ? [
          {
            label: labels.generic,
            component(props): JSX.Element {
              return <AppResourceTextEditor {...props} />;
            },
          },
        ]
      : filterArray(
          Object.entries(editors).map(([type, Editor]) =>
            typeof Editor === 'function'
              ? {
                  label: labels[type],
                  component(props): JSX.Element {
                    return (
                      <>
                        {type === 'visual' && (
                          <OtherCollectionWarning directory={props.directory} />
                        )}
                        <Editor {...props} />
                      </>
                    );
                  },
                }
              : undefined
          )
        );
  }, [subType]);
}

const labels: RR<AppResourceEditorType, LocalizedString> = {
  visual: resourcesText.visualEditor(),
  xml: resourcesText.xmlEditor(),
  json: resourcesText.jsonEditor(),
  generic: resourcesText.textEditor(),
};

/* Display a warning when editing resources from a different collection */
function OtherCollectionWarning({
  directory,
}: {
  readonly directory: SerializedResource<SpAppResourceDir>;
}): JSX.Element | null {
  const isOtherScope = React.useMemo(
    () =>
      (typeof directory.collection === 'string' &&
        directory.collection !==
          getResourceApiUrl('Collection', schema.domainLevelIds.collection)) ||
      (typeof directory.discipline === 'string' &&
        directory.discipline !==
          getResourceApiUrl('Discipline', schema.domainLevelIds.discipline)),
    [directory]
  );
  return isOtherScope ? (
    <WarningMessage>{resourcesText.wrongScopeWarning()}</WarningMessage>
  ) : null;
}

export function Tabs({
  tabs,
  index: [currentIndex, handleChange],
}: {
  readonly tabs: IR<JSX.Element>;
  readonly index: GetSet<number>;
}): JSX.Element {
  return (
    <Tab.Group selectedIndex={currentIndex} onChange={handleChange}>
      <Tab.List
        // Don't display tabs if there is only one tab
        className={`
          inline-flex w-fit flex-wrap gap-2 rounded
          bg-[color:var(--form-background)]
          ${Object.keys(tabs).length === 1 ? 'sr-only' : ''}
        `}
      >
        {Object.keys(tabs).map((label, index) => (
          <Tab
            className={radioButtonClassName(currentIndex === index)}
            key={index}
            /**
             * HeadlessUI does not trigger onChange on click on current tab.
             * This is a workaround. It overrides their click handler only
             * if the option IS current.
             */
            onClick={
              currentIndex === index
                ? (): void => handleChange(index)
                : undefined
            }
          >
            {label}
          </Tab>
        ))}
      </Tab.List>
      <Tab.Panels className="flex flex-1 overflow-hidden">
        {Object.values(tabs).map((element, index) => (
          <Tab.Panel className="flex flex-1 flex-col gap-4" key={index}>
            <ErrorBoundary dismissible>{element}</ErrorBoundary>
          </Tab.Panel>
        ))}
      </Tab.Panels>
    </Tab.Group>
  );
}
