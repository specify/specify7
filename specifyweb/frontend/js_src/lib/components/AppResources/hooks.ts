import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { useAsyncState } from '../../hooks/useAsyncState';
import { f } from '../../utils/functools';
import type { GetOrSet, IR, RA } from '../../utils/types';
import { addMissingFields } from '../DataModel/addMissingFields';
import { fetchCollection } from '../DataModel/collection';
import type { SerializedResource } from '../DataModel/helperTypes';
import type {
  Collection,
  Discipline,
  SpAppResource,
  SpAppResourceData,
  SpecifyUser,
  SpViewSetObj,
} from '../DataModel/types';
import { userPreferences } from '../Preferences/userPreferences';
import { getAppResourceType } from './filtersHelpers';
import { getAppResourceCount, getAppResourceMode } from './helpers';
import { getAppResourceTree, getScope } from './tree';
import type { ScopedAppResourceDir } from './types';
import { appResourceSubTypes } from './types';

export type AppResources = {
  readonly directories: RA<ScopedAppResourceDir>;
  readonly disciplines: RA<SerializedResource<Discipline>>;
  readonly collections: RA<SerializedResource<Collection>>;
  readonly users: RA<SerializedResource<SpecifyUser>>;
  readonly appResources: RA<SerializedResource<SpAppResource>>;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
};

export function useAppResources(
  loadingScreen: boolean = true
): GetOrSet<AppResources | undefined> {
  return useAsyncState(
    React.useCallback(
      async () =>
        f.all({
          directories: fetchCollection('SpAppResourceDir', {
            limit: 0,
            domainFilter: false,
          }).then<AppResources['directories']>(({ records }) =>
            records.map((record) => ({ ...record, scope: getScope(record) }))
          ),
          disciplines: fetchCollection('Discipline', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          collections: fetchCollection('Collection', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          users: fetchCollection('SpecifyUser', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          appResources: fetchCollection('SpAppResource', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          viewSets: fetchCollection('SpViewSetObj', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
        }),
      []
    ),
    loadingScreen
  );
}

export type AppResourcesTree = RA<{
  readonly label: LocalizedString;
  /*
   * A string that would be stable thought the lifecycle of an object.
   * Used to identify a tree node when storing conformation it in localStorage.
   *
   * Directory key is created instead of using id, because app resource
   * needs a way to identify directories that don't yet have IDs (i.e, when
   * creating an app resource for a collection that didn't have any app resources
   * before, the collection won't have any SpAppResourceDir record, thus not
   * directory ID).
   */
  readonly key: string;
  readonly directory: ScopedAppResourceDir | undefined;
  readonly appResources: RA<
    SerializedResource<SpAppResource> & {
      readonly label?: LocalizedString;
    }
  >;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
  readonly subCategories: AppResourcesTree;
}>;

export function useResourcesTree(resources: AppResources): AppResourcesTree {
  const [localize] = userPreferences.use(
    'appResources',
    'appearance',
    'localizeResourceNames'
  );
  return React.useMemo<AppResourcesTree>(() => {
    const tree = getAppResourceTree(resources);
    return localize ? localizeTree(tree) : tree;
  }, [resources, localize]);
}

const localizeTree = (tree: AppResourcesTree): AppResourcesTree =>
  tree.map(({ appResources, subCategories, ...rest }) => ({
    ...rest,
    appResources: appResources.map(localizeResource),
    subCategories: localizeTree(subCategories),
  }));

function localizeResource(
  resource: SerializedResource<SpAppResource> & {
    readonly label?: LocalizedString;
  }
): SerializedResource<SpAppResource> & { readonly label?: LocalizedString } {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  // Check that resource of this type can only have one specific name
  return typeof type.name === 'string'
    ? {
        ...resource,
        // Then replace the name with a localized label unless it's already set
        label: resource.label ?? type.label,
      }
    : resource;
}

export function useAppResourceCount(
  resourcesTree: AppResourcesTree[number]
): number {
  return React.useMemo(
    () => getAppResourceCount(resourcesTree),
    [resourcesTree]
  );
}

/**
 * Fetch resource contents
 */
export function useAppResourceData(
  resource: SerializedResource<SpAppResource | SpViewSetObj>,
  initialData: string | undefined
): {
  readonly resourceData: GetOrSet<
    SerializedResource<SpAppResourceData> | undefined
  >;
  readonly isChanged: boolean;
} {
  const initialValue = React.useRef<string>('');
  const [resourceData, setResourceData] = useAsyncState<
    SerializedResource<SpAppResourceData>
  >(
    React.useCallback(async () => {
      const relationshipName =
        getAppResourceMode(resource) === 'appResources'
          ? 'spAppResource'
          : 'spViewSetObj';
      const newResource = addMissingFields('SpAppResourceData', {
        [relationshipName]: resource.id,
        data: initialData,
      });
      const dataResource =
        typeof resource.id === 'number'
          ? await fetchCollection('SpAppResourceData', {
              limit: 1,
              [relationshipName]: resource.id,
              domainFilter: false,
            }).then(
              ({ records }) =>
                /*
                 * For some reason, app resource can have multiple app resource
                 * datas (but it never does in practice)
                 */
                records[0] ?? newResource
            )
          : newResource;
      const newData = fixLineBreaks(dataResource.data ?? '');
      initialValue.current = newData;
      return { ...dataResource, data: newData };
    }, [resource, initialData]),
    true
  );
  return {
    resourceData: [resourceData, setResourceData],
    isChanged: initialValue.current !== resourceData?.data,
  };
}

const fixLineBreaks = (string: string): string =>
  string.replaceAll(/[\n\r]+/gu, '\n');

export const getAppResourceExtension = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): string =>
  resource._tableName === 'SpViewSetObj'
    ? 'xml'
    : getResourceExtension(resource as SerializedResource<SpAppResource>);

function getResourceExtension(
  resource: SerializedResource<SpAppResource>
): 'jrxml' | 'json' | 'properties' | 'txt' | 'xml' {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  const mimeType = resource.mimeType?.toLowerCase() ?? type?.mimeType ?? '';
  if (mimeType in mimeMapper) return mimeMapper[mimeType];
  else if (mimeType.startsWith('jrxml')) return 'jrxml';
  else if (
    resource.name === 'preferences' &&
    mimeType === appResourceSubTypes.otherPropertiesResource.mimeType
  )
    return 'properties';
  else return 'txt';
}

const mimeMapper: IR<'json' | 'properties' | 'txt' | 'xml'> = {
  'application/json': 'json',
  'text/xml': 'xml',
  'text/plain': 'txt',
};
