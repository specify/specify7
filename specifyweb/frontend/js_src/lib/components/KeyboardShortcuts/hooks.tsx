import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { localized } from '../../utils/types';
import { formatDisjunction } from '../Atoms/Internationalization';
import type { KeyboardShortcuts } from './config';
import { bindKeyboardShortcut } from './context';
import { localizeKeyboardShortcut, resolvePlatformShortcuts } from './utils';

/**
 * Prefer `userPreferences.useKeyboardShortcut()` over directly using this hook
 */
export function useManualKeyboardShortcut(
  shortcuts: KeyboardShortcuts | undefined,
  callback: (() => void) | undefined
): void {
  const callbackRef = React.useRef(callback);
  callbackRef.current = callback;
  const hasCallback = typeof callback === 'function';
  React.useEffect(
    () =>
      typeof shortcuts === 'object' && hasCallback
        ? bindKeyboardShortcut(shortcuts, () => callbackRef.current?.())
        : undefined,
    [hasCallback, shortcuts]
  );
}

/**
 * Prefer `userPreferences.useKeyboardShortcut()` over directly using this hook
 *
 * Provides a localized keyboard shortcut string, which can be used in the UI
 * in the "title" attribute.
 */
export function useKeyboardShortcutLabel(
  shortcuts: KeyboardShortcuts | undefined
): LocalizedString {
  return React.useMemo(() => {
    const platformShortcuts =
      shortcuts === undefined ? [] : resolvePlatformShortcuts(shortcuts) ?? [];
    return localized(
      platformShortcuts.length > 0
        ? ` (${formatDisjunction(
            platformShortcuts.map(localizeKeyboardShortcut)
          )})`
        : ''
    );
  }, [shortcuts]);
}
