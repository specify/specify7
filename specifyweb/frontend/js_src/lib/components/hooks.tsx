import React from 'react';

import type { AnySchema } from '../datamodelutils';
import type { SpecifyResource } from '../legacytypes';
import commonText from '../localization/common';
import * as navigation from '../navigation';
import type { Input } from '../saveblockers';
import type { R, RA } from '../types';
import type { Parser } from '../uiparse';
import { parseValue, resolveParser } from '../uiparse';
import { isInputTouched } from '../validationmessages';
import { LoadingContext } from './contexts';
import { f } from '../functools';

const idStore: R<number> = {};

/**
 * A hook that returns a unique string ID generator that is unique
 * and unchanging for the lifecycle of a component
 */
export function useId(prefix: string): (suffix: string) => string {
  const id = React.useRef(-1);

  const resolvedPrefix = `${prefix}-`;

  if (!(resolvedPrefix in idStore)) idStore[resolvedPrefix] = 0;

  if (id.current === -1) {
    id.current = idStore[resolvedPrefix];
    idStore[resolvedPrefix] += 1;
  }

  return React.useCallback(
    (suffix = ''): string =>
      `${resolvedPrefix}${id.current}${suffix ? `-${suffix}` : ''}`,
    [resolvedPrefix]
  );
}

export function setTitle(title: string): void {
  window.document.title = commonText('appTitle')(title);
}

/** Set title of the webpage. Restores previous title on component destruction */
export function useTitle(title: string): void {
  // Reset title after component is destroyed
  React.useEffect(() => {
    const initialTitle = document.title;
    return (): void => {
      document.title = initialTitle;
    };
  }, []);

  // Change page's title
  React.useEffect(() => {
    setTitle(title);
  }, [title]);
}

/**
 * An integration into native browser error reporting mechanism.
 * Can set an error message via prop or callback.
 * Hides the error message on input
 *
 * @remarks
 * For performance reasons, this hook does not cause state update when setting
 * validation message. Thus, you can call it on keydown to implement live
 * validation
 */
export function useValidation<T extends Input = HTMLInputElement>(
  // Can set validation message from state or a prop
  message: string | RA<string> = ''
): {
  // Set this as a ref prop on an input
  readonly validationRef: React.RefCallback<T>;
  // If need access to the underlying inputRef, can use this prop
  readonly inputRef: React.MutableRefObject<T | null>;
  // Can set validation message via this callback
  readonly setValidation: (message: string | RA<string>) => void;
} {
  const inputRef = React.useRef<T | null>(null);

  /*
   * Store last validation message in case inputRef.current is null at the moment
   * This happens if setValidation is called for an input that is not currently
   * rendered
   */
  const validationMessageRef = React.useRef<string>(
    Array.isArray(message) ? message.join('\n') : message
  );

  // Clear validation message on typing
  React.useEffect(() => {
    if (!inputRef.current) return undefined;
    const input = inputRef.current;

    function handleChange(): void {
      if (input.validity.customError) {
        validationMessageRef.current = '';
        input.setCustomValidity('');
      }
    }

    input.addEventListener('input', handleChange);
    return (): void => input.removeEventListener('input', handleChange);
  }, []);

  const setValidation = React.useCallback(function setValidation(
    message: string | RA<string>
  ): void {
    const joined = Array.isArray(message) ? message.join('\n') : message;
    if (validationMessageRef.current === joined) return;

    validationMessageRef.current = joined;
    const input = inputRef.current;
    if (!input) return;
    // Empty string clears validation error
    input.setCustomValidity(joined);

    if (joined !== '' && isInputTouched(input)) input.reportValidity();
  },
  []);

  React.useEffect(() => setValidation(message), [message, setValidation]);

  return {
    inputRef,
    validationRef: React.useCallback(
      (input): void => {
        inputRef.current = input;
        setValidation(validationMessageRef.current);
      },
      [setValidation]
    ),
    setValidation,
  };
}

/**
 * Like React.useState, but initial value is retrieved asynchronously
 * While value is being retrieved, hook returns undefined, which can be
 * conveniently replaced with a default value when destructuring the array
 *
 * @remarks
 * This hook resets the state value every time the prop changes. Thus,
 * you need to wrap the prop in React.useCallback(). This allows for
 * recalculation of the state when parent component props change.
 *
 * If async action is resolved after component destruction, no update occurs
 * (thus no warning messages are triggered)
 *
 * Rejected promises result in a modal error dialog
 *
 * @example
 * This would fetch data from a url, use defaultValue while fetching,
 * reFetch every time url changes, and allow to manually change state
 * value using setValue:
 * ```js
 * const [value=defaultValue, setValue] = useAsyncState(
 *   React.useCallback(()=>fetch(url), [url]);
 * );
 * ```
 */
export function useAsyncState<T>(
  callback: () => undefined | T | Promise<T | undefined>,
  // Show the loading screen while the promise is being resolved
  loadingScreen: boolean,
  // Whether to reset state value to undefined while fetching new state
  forceConsistency = false
): [
  state: T | undefined,
  setState: React.Dispatch<React.SetStateAction<T | undefined>>
] {
  const [state, setState] = React.useState<T | undefined>(undefined);
  const loading = React.useContext(LoadingContext);

  React.useEffect(() => {
    if (forceConsistency) setState(undefined);
    const wrapped = loadingScreen ? loading : f.id;
    void wrapped(
      Promise.resolve(callback()).then((newState) =>
        destructorCalled ? undefined : setState(newState)
      )
    );

    let destructorCalled = false;
    return (): void => {
      destructorCalled = true;
    };
  }, [callback, loading, loadingScreen]);

  return [state, setState];
}

/**
 * A synchronous version of useAsyncState
 *
 * @remarks
 * Like React.useState, but default value must always be a function, and when
 * function changes, default value is recalculated and reapplied.
 *
 * Thus, wrap the callback in React.useCallback with dependency array that
 * would determine when the state is recalculated.
 *
 * @example
 * This will call getDefaultValue to get new default value every time
 * dependency changes
 * ```js
 * const [value, setValue] = useLiveState(
 *   React.useCallback(
 *     getDefaultValue,
 *     [dependency]
 *   )
 * );
 * ```
 */
export function useLiveState<T>(
  callback: () => T
): [state: T, setState: React.Dispatch<React.SetStateAction<T>>] {
  const [state, setState] = React.useState<T>(() => callback());

  useReadyEffect(React.useCallback(() => setState(callback()), [callback]));

  return [state, setState];
}

/**
 * Like React.useState, but updates the state whenever default value changes
 */
export function useTriggerState<T>(
  defaultValue: T
): [state: T, setState: React.Dispatch<React.SetStateAction<T>>] {
  const [state, setState] = React.useState<T>(defaultValue);

  /* Using layout effect rather than useEffect to update the state earlier on */
  React.useLayoutEffect(() => {
    setState(defaultValue);
  }, [defaultValue]);

  return [state, setState];
}

/**
 * Like React.useEffect, but does not execute on first render.
 * Passed callback must be wrapped in React.useCallback
 */
export function useReadyEffect(callback: () => void): void {
  const isFirstRender = React.useRef(true);
  React.useEffect(() => {
    if (isFirstRender.current) isFirstRender.current = false;
    else callback();
  }, [callback]);
}

export function useUnloadProtect(
  isEnabled: boolean,
  message: string
): (isEnabled: boolean, callback?: () => void) => void {
  const [hasUnloadProtect, setHasUnloadProtect] = useTriggerState(isEnabled);

  React.useEffect(() => {
    if (!hasUnloadProtect) return undefined;
    console.log(`Adding unload protect: ${message}`);
    const id = {};
    navigation.addUnloadProtect(id, message);
    return (): void => {
      console.log(`Removing unload protect: ${message}`);
      navigation.removeUnloadProtect(id);
    };
  }, [hasUnloadProtect, message]);

  const [callback, setCallback] = React.useState<(() => void) | undefined>(
    undefined
  );
  if (typeof callback === 'function') {
    callback();
    setCallback(undefined);
  }
  return React.useCallback((isEnabled, callback) => {
    setHasUnloadProtect(isEnabled);
    setCallback(callback);
  }, []);
}

/**
 * Many react states are simple boolean switches
 * This hook gives a convenient way to defined such states
 *
 * @example Usage
 * Without this hook:
 * ```js
 * const [isOpen, setIsOpen] = React.useState(false);
 * ```
 * With this hook:
 * ```
 * const [isOpen, handleOpen, handleClose, handleToggle] = useBooleanState();
 * ```
 * "handleOpen" is easier to reason about than "setIsOpen(false)"
 *
 * If handleClose or handleToggle actions are not needed, they simply
 * don't have to be destructured.
 *
 * Initial value can be given as a prop. State value is changed to match the
 * prop if prop changes.
 *
 * @example Performance optimization
 * This hook also reduces the render the need for reRenders
 * This calls reRender of Dialog on each parent component render since
 * lamda function is redefined at each render:
 * ```js
 * <Dialog onClose={():void => setIsOpen(false)} ... >...</Dialog>
 * ```
 * This doss not cause needless reRenders and looks cleaner:
 * ```js
 * <Dialog onClose={handleClose} ... >...</Dialog>
 * ```
 */
export function useBooleanState(
  value = false
): Readonly<
  [state: boolean, enable: () => void, disable: () => void, toggle: () => void]
> {
  const [state, setState] = useTriggerState(value);
  return [
    state,
    React.useCallback(
      function enable() {
        setState(true);
      },
      [setState]
    ),
    React.useCallback(
      function disable() {
        setState(false);
      },
      [setState]
    ),
    React.useCallback(
      function toggle() {
        setState((value) => !value);
      },
      [setState]
    ),
  ];
}

/**
 * A hook to integrate an Input with a field on a Backbone resource
 *
 * @remarks
 * If Backbone field value changes, hook is updated
 *
 * Field schema is used to define a Parser than can be used to get
 * validation attributes for an Input
 *
 * If field value is invalid, save blocker is set. It is cleared as soon
 * as field value is corrected
 *
 */
export function useResourceValue<
  T extends string | number | boolean,
  INPUT extends Input = HTMLInputElement
>(
  resource: SpecifyResource<AnySchema>,
  // If fieldName is undefined, this behaves pretty much like useValidation()
  fieldName: string | undefined,
  defaultParser: Parser | undefined,
  validationMessage?: string | RA<string>
): {
  readonly value: T | undefined;
  readonly updateValue: (newValue: T) => void;
  // See useValidation for documentation of these props:
  readonly validationRef: React.RefCallback<INPUT>;
  readonly inputRef: React.MutableRefObject<INPUT | null>;
  readonly setValidation: (message: string | RA<string>) => void;
  readonly parser: Parser;
} & ReturnType<typeof useValidation> {
  const { inputRef, validationRef, setValidation } =
    useValidation<INPUT>(validationMessage);

  const [parser, setParser] = React.useState<Parser>({});

  const [value, setValue] = React.useState<T | undefined>(undefined);

  // Parse value and update saveBlockers
  const updateValue = React.useCallback(
    function updateValue(newValue: T) {
      setValue(
        (typeof parser?.type === 'number'
          ? f.parseInt(parser?.printFormatter?.(newValue, parser) ?? '') ??
            newValue
          : newValue) as T
      );
      if (typeof fieldName !== 'string') return;

      const parseResults = parseValue(
        parser,
        inputRef.current ?? undefined,
        newValue?.toString() ?? ''
      );
      const key = `parseError:${fieldName.toLowerCase()}`;
      if (parseResults.isValid) {
        resource.saveBlockers?.remove(key);
        const oldValue = resource.get(fieldName) ?? null;
        const parsedValue = parseResults.parsed as string;
        // Don't trigger unload protect needlessly
        if (oldValue !== parsedValue) {
          resource.set(fieldName, parsedValue as never);
        }
      } else {
        setValidation(parseResults.reason);
        resource.saveBlockers?.add(key, fieldName, parseResults.reason);
      }
    },
    [resource, fieldName, parser, inputRef, setValidation]
  );

  /*
   * Show errors if default parser changes (to catch possible performance issues)
   */
  const previousParser = React.useRef<Parser | undefined | false>(false);
  React.useEffect(() => {
    if (
      previousParser.current !== false &&
      previousParser.current !== defaultParser
    )
      console.error('Default parser changed. Use React.useMemo()');
    return (): void => {
      previousParser.current = defaultParser;
    };
  }, [defaultParser]);

  // Listen for resource update. Set parser
  React.useEffect(() => {
    if (typeof fieldName === 'undefined') return undefined;

    const field = resource.specifyModel.getField(fieldName);
    if (typeof field === 'undefined')
      console.error(
        `${fieldName} does not exist on ${resource.specifyModel.name}`,
        { resource }
      );

    setParser(
      resource.noValidation === true || typeof field === 'undefined'
        ? {}
        : resolveParser(field) ?? {}
    );

    resource.settingDefaultValues(() =>
      typeof defaultParser?.value === 'undefined' || !resource.isNew()
        ? undefined
        : resource.set(fieldName, defaultParser.value as never)
    );

    const refresh = (): void =>
      setValue((resource.get(fieldName) as T | null) ?? undefined);

    resource.on(`change:${fieldName}`, refresh);
    refresh();
    return (): void => resource.off(`change:${fieldName}`, refresh);
  }, [resource, fieldName, defaultParser]);

  return {
    value,
    updateValue,
    inputRef,
    validationRef,
    setValidation,
    parser,
  } as const;
}
