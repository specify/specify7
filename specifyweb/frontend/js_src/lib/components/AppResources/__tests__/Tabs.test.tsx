import React from 'react';

import { clearIdStore } from '../../../hooks/useId';
import { mount } from '../../../tests/reactUtils';
import { Tabs } from '../Tabs';

beforeEach(() => {
  clearIdStore();
});

describe('Tabs', () => {
  test('simple no tabs', () => {
    const handleChange = jest.fn();

    const { asFragment } = mount(<Tabs index={[0, handleChange]} tabs={{}} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('first tab selected, and then second', async () => {
    const handleChange = jest.fn();

    const tabs = {
      tab1: <h1>Tab1</h1>,
      tab2: <h1>Tab2</h1>,
    };

    const { asFragment, getAllByRole, user } = mount(
      <Tabs index={[0, handleChange]} tabs={tabs} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(handleChange).not.toHaveBeenCalled();

    const tabElements = getAllByRole('tab');
    await user.click(tabElements[1]);

    expect(handleChange).toHaveBeenCalled();
    expect(handleChange).toHaveBeenCalledWith(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
