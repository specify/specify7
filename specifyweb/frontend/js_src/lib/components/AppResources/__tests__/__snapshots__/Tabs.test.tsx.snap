// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs first tab selected, and then second 1`] = `
<DocumentFragment>
  <div
    aria-orientation="horizontal"
    class="
          inline-flex w-fit flex-wrap gap-2 rounded
          bg-[color:var(--form-background)]
          
        "
    role="tablist"
  >
    <button
      aria-controls="headlessui-tabs-panel-:r2:"
      aria-selected="true"
      class="
  button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm aria-handled
  dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white
"
      id="headlessui-tabs-tab-:r0:"
      role="tab"
      tabindex="0"
      type="button"
    >
      tab1
    </button>
    <button
      aria-selected="false"
      class="
  button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm aria-handled
  hover:bg-gray-300 hover:dark:bg-neutral-600
"
      id="headlessui-tabs-tab-:r1:"
      role="tab"
      tabindex="-1"
      type="button"
    >
      tab2
    </button>
  </div>
  <div
    class="flex flex-1 overflow-hidden"
  >
    <div
      aria-labelledby="headlessui-tabs-tab-:r0:"
      class="flex flex-1 flex-col gap-4"
      id="headlessui-tabs-panel-:r2:"
      role="tabpanel"
      tabindex="0"
    >
      <h1>
        Tab1
      </h1>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs first tab selected, and then second 2`] = `
<DocumentFragment>
  <div
    aria-orientation="horizontal"
    class="
          inline-flex w-fit flex-wrap gap-2 rounded
          bg-[color:var(--form-background)]
          
        "
    role="tablist"
  >
    <button
      aria-selected="false"
      class="
  button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm aria-handled
  dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white
"
      id="headlessui-tabs-tab-:r0:"
      role="tab"
      tabindex="-1"
      type="button"
    >
      tab1
    </button>
    <button
      aria-controls="headlessui-tabs-panel-:r3:"
      aria-selected="true"
      class="
  button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm aria-handled
  hover:bg-gray-300 hover:dark:bg-neutral-600
"
      id="headlessui-tabs-tab-:r1:"
      role="tab"
      tabindex="0"
      type="button"
    >
      tab2
    </button>
  </div>
  <div
    class="flex flex-1 overflow-hidden"
  >
    <div
      aria-labelledby="headlessui-tabs-tab-:r1:"
      class="flex flex-1 flex-col gap-4"
      id="headlessui-tabs-panel-:r3:"
      role="tabpanel"
      tabindex="0"
    >
      <h1>
        Tab2
      </h1>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs simple no tabs 1`] = `
<DocumentFragment>
  <button
    aria-hidden="true"
    style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
    type="button"
  />
  <div
    aria-orientation="horizontal"
    class="
          inline-flex w-fit flex-wrap gap-2 rounded
          bg-[color:var(--form-background)]
          
        "
    role="tablist"
  />
  <div
    class="flex flex-1 overflow-hidden"
  />
</DocumentFragment>
`;
